### PR TITLE
Feedback

### PR DESCRIPTION
**Pull request by bitfield**
:wave:! GitHub Classroom created this pull request as a place for your teacher to leave feedback on your work. It will update automatically. **Don’t close or merge this pull request**, unless you’re instructed to do so by your teacher.

In this pull request, your teacher can leave comments and feedback on your code. Click the **Subscribe** button to be notified if that happens.

Click the **Files changed** or **Commits** tab to see all of the changes pushed to `master` since the assignment started. Your teacher can see this too.

<details>
<summary>
<strong>Notes for teachers</strong>
</summary>

Use this PR to leave feedback. Here are some tips:

- Click the **Files changed** tab to see all of the changes pushed to `master` since the assignment started. To leave comments on specific lines of code, put your cursor over a line of code and click the blue **+** (plus sign). To learn more about comments, read “[Commenting on a pull request](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request)”.
- Click the **Commits** tab to see the commits pushed to `master`. Click a commit to see specific changes.
- If you turned on autograding, then click the **Checks** tab to see the results.
- This page is an overview. It shows commits, line comments, and general comments. You can leave a general comment below.

For more information about this pull request, read “[Leaving assignment feedback in GitHub](/help/leaving-feedback-in-github)”.
</details>

Subscribed: @vanjara

### CODE REVIEW

File: 08/calculator.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package calculator
+
+func Add(a int, b int) int {
```
Comment (bitfield) [inline 08/calculator.go]: Just so you know, when you have a series of parameters of the same type, you can declare them like this:

```suggestion
func Add(a, b int) int {
```

File: 09/calculator_test.go [hunk 2]
```diff
@@ -1,17 +1,57 @@
...
 
 func TestSubtract(t *testing.T) {
-	t.Fatal("Not implemented yet")
+	testCases := []struct {
+		a, b int
+		want int
+	}{
+		{a: 1, b: 1, want: 0},
+		{a: 3, b: 2, want: 1},
+		{a: 6, b: 7, want: -1},
```
Comment (bitfield) [inline 09/calculator_test.go:15]: Another interesting test case might be if we subtract a negative number from a positive one.

File: 09/calculator_test.go [hunk 3]
```diff
@@ -1,17 +1,57 @@
...
+	}
+	for _, testCase := range testCases {
+		got := calculator.Subtract(testCase.a, testCase.b)
+		if testCase.want != got {
+			t.Errorf("Add(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
```
Comment (bitfield) [inline 09/calculator_test.go]:
```suggestion
			t.Errorf("Subtract(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
```

File: 09/calculator_test.go [hunk 4]
```diff
@@ -1,17 +1,57 @@
...
+		want int
+	}{
+		{a: 1, b: 1, want: 1},
+		{a: 3, b: 2, want: 6},
+		{a: 6, b: 7, want: 42},
+	}
+	for _, testCase := range testCases {
+		got := calculator.Multiply(testCase.a, testCase.b)
+		if testCase.want != got {
+			t.Errorf("Add(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
```
Comment (bitfield) [inline 09/calculator_test.go]:
```suggestion
			t.Errorf("Multiply(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
```

File: 09/calculator_test.go [hunk 5]
```diff
@@ -1,17 +1,57 @@
...
+		want int
+	}{
+		{a: 1, b: 1, want: 1},
+		{a: 4, b: 2, want: 2},
+		{a: 42, b: 7, want: 6},
+	}
+	for _, testCase := range testCases {
+		got := calculator.Divide(testCase.a, testCase.b)
+		if testCase.want != got {
+			t.Errorf("Add(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
```
Comment (bitfield) [inline 09/calculator_test.go]:
```suggestion
			t.Errorf("Divide(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
```

File: 10/calculator_test.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	}
+	for _, testCase := range testCases {
+		got := calculator.Subtract(testCase.a, testCase.b)
+		if testCase.want != got {
+			t.Errorf("Add(%d, %d): want %d, got %d", testCase.a, testCase.b, testCase.want, got)
+		}
+	}
+}
+
+func TestDivide(t *testing.T) {
```
Comment (bitfield) [inline 10/calculator_test.go]: It might make sense for us to move `TestDivide` so that the other three tests (which are virtually identical in structure) are grouped together.

File: 10/calculator_test.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	testCases := []struct {
+		a, b          int
+		want          int
+		errorExpected bool
+	}{
+		{a: 1, b: 1, want: 1, errorExpected: false},
+		{a: 4, b: 0, want: 2, errorExpected: true},
```
Comment (bitfield) [inline 10/calculator_test.go]:
```suggestion
		{a: 4, b: 0, want: 0, errorExpected: true},
```

Since `errorExpected: true` signals that we won't be checking the data value anyway, it doesn't matter what `want` we use here, but it makes sense to use zero, in the same way that we return zero from functions in the error case.

File: 10/calculator_test.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+		{a: 42, b: 7, want: 6, errorExpected: false},
+	}
+	for _, testCase := range testCases {
+		got, err := calculator.Divide(testCase.a, testCase.b)
+		if err == nil {
```
Comment (bitfield) [inline 10/calculator_test.go]: There are actually four possible outcomes here, aren't there, each of which the test needs to handle:

1. We didn't expect an error, and we didn't get one, and `got == want`—succeed
2. We didn't expect an error, and we didn't get one, but `got != want`—fail
3. We didn't expect an error, but we did get one—fail
4. We _did_ expect an error, but we _didn't_ get one—fail

Try modifying either the code or the test to trigger each of these possible outcomes, and see if the test does the right thing in that case. (By 'the right thing', what I really mean is not only that the test should pass or fail as necessary, but that if it fails, it accurately explains what was wrong.)

**Review by bitfield — COMMENTED** [review]
Looking very good!

**Comment by bitfield** [comment]
Hey @vanjara, did you see my feedback here? I was about to comment on the other PR, but realised I'd be saying the same thing again 😁

File: 08/calculator.go (hunk 1 above)
Comment (vanjara) [reply 08/calculator.go]: Thank you for the suggestion.

File: 09/calculator_test.go (hunk 2 above)
Comment (vanjara) [reply 09/calculator_test.go:15]: Thanks, added this test case.
